src/python/pants/backend/nfpm/field_sets.py

```diff
@@ -0,0 +1,87 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from abc import ABCMeta
+from dataclasses import dataclass
+from typing import Any, ClassVar, cast
+
+from typing_extensions import Protocol
+
+from pants.backend.nfpm.fields.all import NfpmPackageNameField
+from pants.backend.nfpm.target_types import APK_FIELDS, ARCHLINUX_FIELDS, DEB_FIELDS, RPM_FIELDS
+from pants.core.goals.package import OutputPathField, PackageFieldSet
+from pants.engine.rules import collect_rules
+from pants.engine.target import DescriptionField, Target
+from pants.engine.unions import UnionRule
+
+
+# This allows us to define the nFPM config option name on the field
+class _NfpmField(Protocol):
+    # nfpm_alias is a "." concatenated series of dict keys (keys of nfpm.yaml).
+    nfpm_alias: ClassVar[str]
+
+
+@dataclass(frozen=True)
+class NfpmPackageFieldSet(PackageFieldSet, metaclass=ABCMeta):
+    output_path: OutputPathField
+    package_name: NfpmPackageNameField
+    description: DescriptionField
+
+    def nfpm_config(self, tgt: Target) -> dict[str, Any]:
+        config: dict[str, Any] = {}
+        for field in self.required_fields:
+            value = tgt[field].value
+            if not field.required and value is None:
+                # omit undefined optional values from nfpm.yaml
+                continue
+
+            # handle nested fields (eg: deb.triggers)
+            keys = cast(_NfpmField, field).nfpm_alias.split(".")
```

Comment:
rather than a `cast` here, aren't mypy satisfied if you declare a `required_fields: ClassVar[Collection[_NfpmField]]` in this fieldset class? (or does it complain about the type already being defined in the base class then?)

Reply:
Ooh. I like that idea. I'll try that

Reply:
I could also drop the protocol and do something like:

```python
if not hasattr(field, "nfpm_alias"):
    continue
```

Reply:
I couldn't figure out how to make a type hint for `required_fields` that was compatible with the base class type and added the field. So, I went with `hasattr` and `getattr`.